### DISCUSSION TITLE
What happens with running transaction upon sigterm?

### DISCUSSION THREAD

**Discussion by jaraqueffdc** [sqlalchemy/sqlalchemy#5910] [Usage Questions] [answered]
I have been looking around but I cannot find exactly what I am trying to understand and I am starting to have the feeling that either I am not looking for the right thing or that it is dead simple and I just do not understand :)

We have several ECS microservices that launch transaction queries with locks on the accessed tables. When we issue a new release the services are stopped and each service will receive a SIGTERM signal since that is what `docker stop` does. We have had some cases in which, after a new release, the database was kind of locked for some time. When looking at the database we saw several locks being active but we cannot 100% know which transactions were holding those locks. We know they come from the services but we cannot see if they are new transactions being issued after the restart or if they are old transactions that have not been stopped, unlocking the accessed tables.

We assumed that when the services stopped the transaction will be finished and the lock released but I am not sure how SQLAlchemy handles the sigterm signal.

Is there anything we need to do to handle them or does the sigterm signal trigger the transaction rollback and release the lock? What would be the standard way of dealing with that?

**Additional context**
We have our services running in ECS and the database is in AWS Redshift. The restart triggers a first scheduler service that issues a bunch of queries to start scheduling jobs. This is why we are not sure if the transactions we see are from the scheduler after restart or hanging transactions from the previous execution. I have not seen this happening with a local postgres database but Redshift is a bit different which is why I am trying to understand the sigterm handling process.

Also, we are not relying on ORM, just engine + connections to execute transactions. We use the `engine.begin()` to start the transaction within a context manager but as an utility for our system we have each method decorated with a decorator that ensures there is a transaction open or opens it before running and have our own context manager for this that has the following enter and exit methods.

```python
    def __enter__(self):
        self._transaction_context = self.engine.begin()
        self._connection = self._transaction_context.__enter__()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self._transaction_context.__exit__(exc_type, exc_val, exc_tb)
        self._connection = None
```

We rely at the end on the enter and exit method of the transaction_context that is returned by the `session.begin()` method so we should be ok here but let me know if there is something you think could be wrong.

Thanks a lot in advance!

**Comment by zzzeek** [comment, marked as answer]
hey there -

SQLAlchemy does not set up any SIGTERM handler or other signal handlers.    This is also not something you would see used by database drivers either, which in this case I would assume is the psycopg2 driver.

Database connections at the level of the process are TCP sockets, so when SIGTERM comes to the Python process, assuming you don't have any handlers in there that might be holding it up, those sockets get torn down.   This is all way outside of anything SQLAlchemy or even psycopg2 would be consulted with.     PostgreSQL itself, and I would assume Redshift, would discard such transactions immediately, unless you were using 2pc/XID features which I doubt redshift supports.

I've not worked with redshift directly, however there are several techniques you can use to see what's happening with the database:

1. use netstat on the docker host(s) to look for connections in ESTABLISHED state.
2. on the DB host run queries against it's activity tables; while Postgresql uses [pg_stat_activity](https://www.postgresql.org/docs/9.2/monitoring-stats.html), a quick google shows Redshift will have similar data in [stv_sessions](https://www.sisense.com/blog/helpful-redshift-admin-queries/) and other tables, you'd want to look in those.  these will show the remote IP number and stuff like that so you can see what's actually going on.

**Reply by jaraqueffdc** [reply]
Thanks a lot @zzzeek ! That is more than enough information. I won't do anything at the moment in our services then and if it happens again after a new release I will try to investigate it further in Redshift.

Thanks a lot!

**Reply by CaselIT** [reply]
I'm not sure if Redshift support it, but with PostgreSQL an easy way to tell from where a connection is happening is using the `application_name`. 
In psycopg2 you can set it by passing it in the `connection_args` of `create_engine`: 
`sa.create_engine(..., connect_args={"application_name": 'app-v1'})`

If you add the application version there then it's easy to see if there are old connections active or only new ones
